File: src/main/scala/edu/monash/honours/SpeedCheck.scala
package edu.monash.honours

import org.apache.spark.SparkConf
import org.apache.spark.streaming.{StreamingContext, Seconds}

object SpeedCheck
{
  val SPEED_UPPER_LIMIT = 85
  val SPEED_LOWER_LIMIT = 0

  def main(args: Array[String])
  {
    val sparkConf = new SparkConf().setMaster("local[2]").setAppName("SpeedCheck")
    val ssc = new StreamingContext(sparkConf, Seconds(1))
    val speedStream = ssc.socketTextStream("localhost", 9999)

    speedStream.map(checkSpeed)

    speedStream.print()

    ssc.start()
    ssc.awaitTermination()
  }

  def checkSpeed(speedValue: String): (String, Boolean) = {
    if (speedValue.asInstanceOf[Double] > SPEED_UPPER_LIMIT || speedValue.asInstanceOf[Double] < SPEED_LOWER_LIMIT) {
      (speedValue, false)
    } else {
      (speedValue, true)
    }
  }
}

File: src/main/python/edu/monash/honours/SpeedCheck.py
__author__ = 'poltak'

from pyspark import SparkContext
from pyspark.streaming import StreamingContext

SPEED_UPPER_LIMIT = 85
SPEED_LOWER_LIMIT = 0
SOCKET_ADDRESS = "localhost"
SOCKET_PORT = 9999


def init_spark(name):
    sc = SparkContext("local[2]", name)
    return StreamingContext(sc, 2)


def close_spark(ssc):
    ssc.start()
    ssc.awaitTermination()


def check_speed(speed):
    if speed > SPEED_UPPER_LIMIT or speed < SPEED_LOWER_LIMIT:
        return speed, False
    else:
        return speed, True


def main():
    ssc = init_spark("pyspark-speed-check")
    speed_stream = ssc.socketTextStream(SOCKET_ADDRESS, SOCKET_PORT)

    speed_stream = speed_stream.map(check_speed)

    speed_stream.pprint()

    close_spark(ssc)


if __name__ == '__main__':
    main()


